notes-0225-universes.tex

```diff
@@ -0,0 +1,272 @@
+
+
+\subsection{Function Equivalence}
+A quasi-inverse of a function $\oftype{f}{A \rightarrow B}$
+($\qinv{f}$) includes
+\begin{itemize}
+\item $\oftype{g}{B \rightarrow A}$
+\item $\oftype{\epsilon}{\dfuntype{y}{B}{\idtype{B}{f(g(y))}{y}}}$
+\item $\oftype{\eta}{\dfuntype{x}{A}{\idtype{A}{g(f(x))}{x}}}$
+\end{itemize}
+This may be formulated as a rather large dependent pair type.
+
+For a good definition of $\isequiv{f}$, we want two things:
+\begin{enumerate}
+\item The definition should be logically equivalent to $\qinv{f}$
+  (either one can be derived given the other)
+\item $\isprop{\isequiv{f}}$ \\
+  This requires that $\isequiv{f}$ has at most one element.  The
+  definition of $\qinv{f}$ fails this requirement.
+\end{enumerate}
+
+We give two possible types to fulfill the requirements on
+$\isequiv{f}$:
+\begin{enumerate}
+\item Half Adjoint Equivalence ($\ishae{f}$)
+
+  This includes $g$, $\epsilon$, and $\eta$ as in the quasi-inverse,
+  but it also includes a fourth element $\tau$:
+  \[\oftype{\tau}{\dfuntype{x}{A}{\idtype{}{\ap_f(\eta(x))}{\epsilon(f(x))}}}\]
+  This $\tau$ may be thought of as ensuring that $\eta$ and $\epsilon$
+  are coherent in some way.
+
+  It is clear how to derive $\qinv{f}$ given $\ishae{f}$, since we
+  already have the $g$, $\epsilon$, and $\eta$.  The other direction
+  of the equivalence is much more difficult to prove.
+
+  Rather than $\tau$, we could have defined $\tau'$:
+  \[\oftype{\tau'}{\dpairtype{y}{B}{\idtype{}{\eta(g(y))}{\ap_g(\epsilon(y))}}}\]
+  This definition is dual to the definition of $\tau$.  It is the
+  other ``half'' in the name of the type.
+  %
+  If the definition required both $\tau$ and $\tau'$, it would not
+  satisfy the requirement of $\isprop{\isequiv{f}}$, and we would
+  require a further element to show that $\tau$ and $\tau'$ were
+  coherent.
+
+\item $\isequiv{f} := \dfuntype{y}{B}{\iscontr{\dpairtype{x}{A}{\idtype{}{f(x)}{y}}}}$
+
+  The inner dependent pair type is defining the pre-image of $y$ (in
+  homotopy theory, this is the ``fiber over $y$'' or that ``all fibers
+  are contractible'').  There is only one element in the pre-image,
+  which gives an inverse function.
+\end{enumerate}
+Both of these definitions are acceptable.  In Agda, we prefer half
+adjoint equivalence because it is easier to use.
+
+A difficulty is that proving $\isprop{\isequiv{f}}$ for many
+definitions requires function extensionality, which cannot be defined
+without univalence.  We choose a definition of $\isequiv$ and use it
+to define univalence, then prove $\isprop$ for it assuming univalence.
+
+
+\subsection{Univalence}
+A precise formulation of univalence:
+\begin{itemize}
+\item Define $\oftype{\idtoequiv}{\idtype{\U}{A}{B} \rightarrow A \simeq B}$ \\
+  Axiom:  $\isequiv{\idtoequiv}$ \\
+  $\idtoequiv := \lambda p.\idJ{A}{\textnormal{a proof of \isequiv{id_A}}}{p}$
+
+\item (Strong) Function Extensionality \\
```

Comment:
Perhaps you want to move this out of this subsection?

Reply:
I don't know what you're suggesting could be moved, or to where.

Reply:
I meant that "function extensionality" does not (directly) belong to a section called "Univalence" unless more words are added.